// @wordpress/block-editor/build/components/inserter/category-tabs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("react");
var _i18n = require("@wordpress/i18n");
var _components = require("@wordpress/components");
var _icons = require("@wordpress/icons");
var _lockUnlock = require("../../../lock-unlock");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const {
  Tabs
} = (0, _lockUnlock.unlock)(_components.privateApis);
function CategoryTabs({
  categories,
  selectedCategory,
  onSelectCategory,
  children
}) {
  return (0, _react.createElement)(Tabs, {
    className: "block-editor-inserter__category-tabs",
    selectOnMove: false,
    selectedTabId: selectedCategory ? selectedCategory.name : null,
    orientation: 'vertical',
    onSelect: categoryId => {
      // Pass the full category object
      onSelectCategory(categories.find(category => category.name === categoryId));
    }
  }, (0, _react.createElement)(Tabs.TabList, {
    className: "block-editor-inserter__category-tablist"
  }, categories.map(category => (0, _react.createElement)(Tabs.Tab, {
    key: category.name,
    tabId: category.name,
    className: "block-editor-inserter__category-tab",
    "aria-label": category.label,
    "aria-current": category === selectedCategory ? 'true' : undefined
  }, (0, _react.createElement)(_components.__experimentalHStack, null, (0, _react.createElement)(_components.FlexBlock, null, category.label), (0, _react.createElement)(_icons.Icon, {
    icon: (0, _i18n.isRTL)() ? _icons.chevronLeft : _icons.chevronRight
  }))))), categories.map(category => (0, _react.createElement)(Tabs.TabPanel, {
    key: category.name,
    tabId: category.name,
    focusable: false,
    className: "block-editor-inserter__category-panel"
  }, children)));
}
var _default = exports.default = CategoryTabs;
